### PR TITLE
feat: add ip_addr field to Request object

### PR DESCRIPTION
**Pull request by IdoKendo**
**Description**

This PR contributes towards #524 by adding the client's IP address to the request object.
By having the client's IP address we can later on implement rate limiting per IP address.

An example assert was added to the middleware base routes.

### CODE REVIEW

File: src/server.rs
Comment (JackThomson2) [inline src/server.rs:382]: Should the ip_addr not be set in this function?

Comment (IdoKendo) [reply src/server.rs:382]: You're right! Was kinda planning on doing it and just missed it in the end, will move it.

**Comment by IdoKendo** [comment]
> @IdoKendo , Thanks for the PR. Great work 😄
> 
> I have a question - should we be rate limiting based on the IP address or per client?

Yeah, so for public-facing APIs it's usually limited by IP. If it is a private API and some authentication is in place it should be blocked by auth. So for the rate limit implementation, we'll need to consider both cases.
Not sure where we're standing on authentication for private APIs, but for now I think adding the IP address in the request object is a necessary first step. Also, it could benefit other features in the future as well.

**Comment by sansyrox** [comment]
> Yeah, so for public-facing APIs it's usually limited by IP. If it is a private API and some authentication is in place it should be blocked by auth. So for the rate limit implementation, we'll need to consider both cases.
Not sure where we're standing on authentication for private APIs, but for now I think adding the IP address in the request object is a necessary first step. Also, it could benefit other features in the future as well.

@IdoKendo , perfect then. The authentication in general is still a WIP. So, this will be a good approach.

File: src/types/request.rs
Comment (JackThomson2) [inline src/types/request.rs:17]: Just thinking, would an optional string make more sense here? Rather than giving them an empty string if it's missing? Just a thought. Would also save allocating if we didn't have one

Comment (IdoKendo) [reply src/types/request.rs:17]: That's actually a good idea, how would that translate on the Python side? as just a `None`?

Comment (sansyrox) [reply src/types/request.rs:17]: @IdoKendo , yes. A `None | str`

File: src/types/request.rs
Comment (JackThomson2) [reply src/types/request.rs:103]: Actix returns the requester info as optional https://docs.rs/actix-web/latest/actix_web/struct.HttpRequest.html#method.peer_addr not sure in what cases this is empty but I guess it's best to handle it at least.

Comment (sansyrox) [reply src/types/request.rs:103]: Ah, okay. Fair enough. :D

**Review by sansyrox — COMMENTED** [review]
One last comment @IdoKendo before we can merge this. Can you please write some docs regarding this in the `docs/examples.md` section?

**Comment by IdoKendo** [comment]
> One last comment @IdoKendo before we can merge this. Can you please write some docs regarding this in the `docs/examples.md` section?

No problem, I will add the docs! 😎

**Review by sansyrox — APPROVED** [review]
Great work! LGTM 😄

**Pull request merged**
